plugins/modules/zabbix_user_role.py

```diff
@@ -0,0 +1,183 @@
+
+
+class UserRole(ZabbixBase):
+    def __find_val(self, outval, inval):
+        if outval == str(inval):
+            return True
+        return False
+
+    def __find_list(self, outval, inval):
+        if set(outval) == set(inval):
+            return True
+        return False
+
+    def __find_dict(self, outval, inval):
+        for out in outval:
+            m = True
+            for k, v in inval.items():
+                if out[k] == str(v):
+                    continue
+                else:
+                    m = False
+            if m:
+                break
+        return m
+
+    def is_part_of(self, inp, out):
+        verdict = True
+        for rule, value in inp.items():
+            if not isinstance(value, list):
+                verdict = verdict and self.__find_val(out.get(rule, ''), value)
+            else:
+                if len(value):
+                    if not isinstance(value[0], dict):
+                        verdict = verdict and self.__find_list(out.get(rule, []), value)
+                    else:
+                        for v in value:
+                            verdict = verdict and self.__find_dict(out.get(rule, {}), v)
+                else:
+                    verdict = verdict and self.__find_list(rule, value)
+        return verdict
+
+    def get_user_role(self, name):
+        result = self._zapi.role.get({
+            "output": "extend",
+            "selectRules": "extend",
+            "filter": {"name": name}
+        })
+        return result
+
+
+def main():
+    msg = "No action"
+    changed = False
+
+    argument_spec = zabbix_utils.zabbix_common_argument_spec()
+    argument_spec.update(dict(
+        state=dict(type='str', required=False, default='present'),
+        name=dict(type='str', required=True),
+        type=dict(type='int', required=False, default=1),
+        rules=dict(type='dict', required=True),
+    ))
+
+    # the AnsibleModule object
+    module = AnsibleModule(
+        argument_spec=argument_spec,
+        supports_check_mode=True
+    )
+
+    state = module.params['state']
+    name = module.params['name']
+    type = module.params['type']
+    rules = module.params['rules']
+
+    user_role = UserRole(module)
+
+    result = user_role.get_user_role(name)
+    if result:
+        if len(result) == 1:
+            role = result[0]
+            if role['readonly'] != 1:
+                roleid = role['roleid']
+                if state == 'absent':
+                    result = user_role._zapi.role.delete([f"{roleid}"])
+                    changed = True
+                    msg = "Role deleted"
+                else:
+                    if not user_role.is_part_of(rules, role['rules']):
+                        result = user_role._zapi.role.update({"roleid": roleid, "rules": rules})
+                        changed = True
+                        msg = "Role updated"
```

Comment:
I think this section and whole functions behind that can be replaced with https://github.com/ansible-collections/community.zabbix/blob/10a4040a8a88d50068d3ecc8475f877db8603864/plugins/module_utils/helpers.py#L113 and the function that follows it.

Reply:
As far as the role is concerned, I'm not interested in the difference between conf and Zabbix, but whether what I configured is exactly present like that in Zabbix, the other keys are don't care.